src/crimson/osd/pg.cc

```diff
@@ -676,7 +678,7 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
     reply->set_enoent_reply_versions(peering_state.get_info().last_update,
 				     peering_state.get_info().last_user_version);
     return seastar::make_ready_future<Ref<MOSDOpReply>>(std::move(reply));
```

Comment:
@xxhdx1985126: how to handle a situation when the lambda wants to return an `interruptible_future` instead of `seastar::future`? Is this use case currently supported? Same question about `handle_exception_type_interruptible`.

Reply:
yes, there's a seastar::futurize<interruptible_future> [here](https://github.com/ceph/ceph/pull/37536/commits/2f5889adc1700ddcea16d7b728ad4a4497f2568c#diff-6e851fa4127b0ffe0f35555be5cef63a2c5cb535fb78831180aaa2a065c4ece6R976) to avoid seastar::future<interruptible_future<...., seastar::future<T>>>

Reply:
I asked because of struggling with an issue.

```
[ 87%] Building CXX object src/crimson/osd/CMakeFiles/crimson-osd.dir/pg.cc.o
In file included from /work/ceph-2/src/crimson/common/interruptible_future.h:10,
                 from /work/ceph-2/src/crimson/common/exception.h:11,
                 from /work/ceph-2/src/crimson/common/gated.h:10,
                 from /work/ceph-2/src/crimson/net/Dispatcher.h:21,
                 from /work/ceph-2/src/crimson/net/chained_dispatchers.h:8,
                 from /work/ceph-2/src/crimson/net/SocketConnection.h:21,
                 from /work/ceph-2/src/msg/Message.h:39,
                 from /work/ceph-2/src/messages/MOSDFastDispatchOp.h:7,
                 from /work/ceph-2/src/messages/MOSDRepOpReply.h:19,
                 from /work/ceph-2/src/crimson/osd/pg.h:17,
                 from /work/ceph-2/src/crimson/osd/pg.cc:4:
/work/ceph-2/src/crimson/common/errorator.h: In instantiation of ‘void crimson::maybe_handle_error_t<ErrorVisitorT, FuturatorT>::handle() [with ErrorT = crimson::stateful_error_t<std::error_code>; ErrorVisi
...
/work/ceph-2/src/crimson/osd/pg.cc:681:5:   required from here
/work/ceph-2/src/crimson/common/errorator.h:280:54: error: static assertion failed: return of Error Visitor is not assignable to future
```

when trying to return `PG::interruptible_future<Ref<MOSDOpReply>>` from `handle_exception_type_interruptible` or `OpsExecuter::osd_op_errorator::all_same_way`, or both:

```diff
+PG::interruptible_future<Ref<MOSDOpReply>> PG::handle_failed_op(
+  const std::error_code& e,
+  ObjectContextRef obc,
+  const OpsExecuter& ox,
+  const MOSDOp& m) const
+{
+  // NOP for demo
+  return seastar::make_ready_future<Ref<MOSDOpReply>>();
+}
+
 PG::interruptible_future<Ref<MOSDOpReply>> PG::do_osd_ops(
   Ref<MOSDOp> m,
   ObjectContextRef obc,
@@ -646,7 +655,6 @@ PG::interruptible_future<Ref<MOSDOpReply>> PG::do_osd_ops(
   }).safe_then_interruptible([this,
                 m,
                 obc,
-                ox_deleter = std::move(ox),
                 rvec = op_info.allows_returnvec()] {
     // TODO: should stop at the first op which returns a negative retval,
     //       cmpext uses it for returning the index of first unmatched byte
@@ -665,32 +673,18 @@ PG::interruptible_future<Ref<MOSDOpReply>> PG::do_osd_ops(
       *m,
       obc->obs.oi.soid);
     return seastar::make_ready_future<Ref<MOSDOpReply>>(std::move(reply));
-  }, OpsExecuter::osd_op_errorator::all_same_way([=] (const std::error_code& e) {
-    assert(e.value() > 0);
-    logger().debug(
-      "do_osd_ops: {} - object {} got error code {}, {}",
-      *m,
-      obc->obs.oi.soid,
-      e.value(),
-      e.message());
-    auto reply = make_message<MOSDOpReply>(
-      m.get(), -e.value(), get_osdmap_epoch(), 0, false);
-    reply->set_enoent_reply_versions(peering_state.get_info().last_update,
-                                    peering_state.get_info().last_user_version);
-    return seastar::make_ready_future<Ref<MOSDOpReply>>(std::move(reply));
-  })).handle_exception_type_interruptible([=](const crimson::osd::error& e) {
+  }, OpsExecuter::osd_op_errorator::all_same_way([ox = ox.get(),
+                                                  m,
+                                                  obc,
+                                                  this] (const std::error_code& e) {
+    return handle_failed_op(e, std::move(obc), *ox, *m);
+  })).handle_exception_type_interruptible([ox_deleter = std::move(ox),
+                                           m,
+                                           obc,
+                                           this] (const crimson::osd::error& e) {
     // we need this handler because throwing path which aren't errorated yet.
-    logger().debug(
-      "do_osd_ops: {} - object {} got unhandled exception {} ({})",
-      *m,
-      obc->obs.oi.soid,
-      e.code(),
-      e.what());
-    auto reply = make_message<MOSDOpReply>(
-      m.get(), -e.code().value(), get_osdmap_epoch(), 0, false);
-    reply->set_enoent_reply_versions(peering_state.get_info().last_update,
-                                    peering_state.get_info().last_user_version);
-    return seastar::make_ready_future<Ref<MOSDOpReply>>(std::move(reply));
+    logger().debug("encountered the legacy error handling path!");
+    return seastar::make_ready_future<Ref<MOSDOpReply>>(nullptr);
   });
 }
```

I really appreciate a hint on what's wrong here.

Reply:
@rzarzynski I think this is because crimson::maybe_handle_error_t's futurator_t is determined by ValueFuncT's return type which is seastar::future<Ref<MOSDOpReply>> and is different from ErrorFuncT's return type -- interruptible_future<Ref<MOSDOpReply>> -- in this case.

Reply:
@xxhdx1985126 : Thanks, Xuehan! `make_interruptible` in the optimistic path of `safe_then_interruptible()` helped with the `OpsExecuter::osd_op_errorator::all_same_way` case.

The `handle_exception_type_interruptible()` one looks different. I believe the implementation actually prohibits returning anything other than `seastar::future`. It's because of the early exit:

```cpp
  template <typename Func,
            typename Result = interrupt_futurize_t<
                std::result_of_t<Func(
                  typename seastar::function_traits<Func>::template arg<0>::type)>>>
  [[gnu::always_inline]]
    // ...
      // ...
      if (!fut.failed()) {
        return seastar::make_ready_future<T...>(fut.get());
      } else {
        // ...
      }
    });
  }

Reply:
@rzarzynski Hi, I don't quite follow you. It seems that handle_exception_type_interruptible is only supposed to return seastar::future or interruptible_future. Do you mean this is problematic?:-)